benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,205 @@
+"""Classes for persisting benchmark results."""
+
+import json
+import os
+from datetime import datetime
+from enum import Enum
+from pathlib import Path
+from typing import Protocol
+
+import boto3
+import boto3.session
+from attr import define, field
+from attrs.validators import instance_of
+from git import Repo
+
+from benchmarks import Benchmark, Result
+
+
+class PathStrategy(Enum):
+    """The way a path extension is constructed."""
+
+    Hierarchical = 1
+    """Refers to the hierarchical path construction of folders."""
+
+    Flat = 2
+    """Refers to the flat path construction of a single file name."""
+
+
+@define
+class PathConstructor:
+    """A class to construct the hierarchical path of a result object."""
```

Comment:
Please add a short description here why the existence of this class is necessary, i.e., that S3 is weird.

Reply:
Also, you should briefly describe how paths are being named and what the scheme here is. I believe that this should be part of some sort of still-to-come-somewhen-in-the-future Benchmarking HowTo, but I think the general information should already be included here.

Reply:
This is in particular relevant due to potential folders that might be created/cause trouble.

Reply:
Liked it, found a typo and made that a new comment.